lib/internal/modules/esm/loader.js

```diff
@@ -589,61 +749,115 @@ class ESMLoader {
+      url: resolverFilePath,
+    } = resolvers[hookIndex];
+    let chainFinished = hookIndex === 0;
+    let shortCircuited = false;
+
+    const context = {
+      conditions: DEFAULT_CONDITIONS,
+      importAssertions,
       parentURL,
-    );
+    };
+
+    const nextResolve = async (suppliedSpecifier, ctx = context) => {
+      --hookIndex; // `nextResolve` has been called, so decrement our pointer
+
+      ({
+        fn: resolver,
+        url: resolverFilePath,
+      } = resolvers[hookIndex]);
+
+      if (hookIndex === 0) { chainFinished = true; }
 
-    const conditions = DEFAULT_CONDITIONS;
+      const hookErrIdentifier = `${resolverFilePath} "resolve"`;
 
-    const defaultResolver = this.#resolvers[0];
+      validateString(
+        suppliedSpecifier,
+        `${hookErrIdentifier} nextResolve(specifier)`,
+      ); // non-strings can be coerced to a url string
+
+      validateObject(ctx, `${hookErrIdentifier} nextResolve(, context)`);
+
+      const output = await resolver(suppliedSpecifier, ctx, nextResolve);
+
+      if (output?.shortCircuit === true) { shortCircuited = true; }
+
+      return output;
+    };
 
-    const resolver = this.#resolvers.length === 1 ?
-      defaultResolver :
-      this.#resolvers[1];
     const resolution = await resolver(
       originalSpecifier,
-      {
-        conditions,
-        importAssertions,
-        parentURL,
-      },
-      defaultResolver,
+      context,
+      nextResolve,
     );
 
-    if (typeof resolution !== 'object') {
+    const hookErrIdentifier = `${resolverFilePath} resolve`;
+
+    if (typeof resolution !== 'object') { // [2]
       throw new ERR_INVALID_RETURN_VALUE(
-        'object',
-        'loader resolve',
+        'an object',
+        hookErrIdentifier,
         resolution,
       );
     }
 
-    const { format, url } = resolution;
+    if (resolution?.shortCircuit === true) { shortCircuited = true; }
+
+    if (!chainFinished && !shortCircuited) {
+      throw new ERR_LOADER_CHAIN_INCOMPLETE('resolve', resolverFilePath);
+    }
+
+    const {
+      format,
+      url,
+    } = resolution;
 
     if (
       format != null &&
-      typeof format !== 'string'
+      typeof format !== 'string' // [2]
     ) {
       throw new ERR_INVALID_RETURN_PROPERTY_VALUE(
-        'string',
-        'loader resolve',
+        'a string',
+        hookErrIdentifier,
         'format',
         format,
       );
     }
 
-    if (typeof url !== 'string') { // non-strings can be coerced to a url string
+    if (typeof url !== 'string') {
+      // non-strings can be coerced to a url string
+      // validateString() throws a less-specific error
       throw new ERR_INVALID_RETURN_PROPERTY_VALUE(
-        'string',
-        'loader resolve',
+        'a url string',
+        hookErrIdentifier,
         'url',
         url,
       );
     }
 
-    new URL(url); // Intentionally trigger error if `url` is invalid
+    try {
+      new URL(url);
+    } catch {
```

Comment:
Any reason this is not using the same `!this.moduleMap.has(url)` check as `nextLoad`?

Reply:
I was thinking it wouldn't reach here, but that's only when it starts out. Will add 🙂